File: ios/CareLog/CareLog/Vitals/ViewModels/GlucoseViewModel.swift
import Foundation
import SwiftUI
import Combine

/// ViewModel for glucose logging screen.
@MainActor
final class GlucoseViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var value: String = ""
    @Published var unit: String = "mg/dL"
    @Published var mealTiming: MealTiming?
    @Published var valueError: String?
    @Published var isSaving: Bool = false
    @Published var saveError: String?

    // MARK: - Computed Properties

    var canSave: Bool {
        guard let doubleValue = Double(value) else { return false }
        return validateValue(doubleValue) == nil
    }

    // MARK: - Dependencies

    private let authService: AuthService
    private let localFHIRRepository: LocalFHIRRepository
    private let voicePlayer: VoiceAcknowledgementPlayer
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init(
        authService: AuthService = .shared,
        localFHIRRepository: LocalFHIRRepository = .shared,
        voicePlayer: VoiceAcknowledgementPlayer = .shared
    ) {
        self.authService = authService
        self.localFHIRRepository = localFHIRRepository
        self.voicePlayer = voicePlayer

        setupValidation()
    }

    // MARK: - Validation

    private func setupValidation() {
        $value
            .dropFirst()
            .sink { [weak self] value in
                guard let self = self,
                      let doubleValue = Double(value) else {
                    self?.valueError = nil
                    return
                }
                self.valueError = self.validateValue(doubleValue)
            }
            .store(in: &cancellables)
    }

    private func validateValue(_ value: Double) -> String? {
        switch unit {
        case "mg/dL":
            switch value {
            case ..<20: return "Too low (min 20)"
            case 601...: return "Too high (max 600)"
            default: return nil
            }
        case "mmol/L":
            switch value {
            case ..<1.1: return "Too low (min 1.1)"
            case 33.4...: return "Too high (max 33.3)"
            default: return nil
            }
        default:
            return nil
        }
    }

    // MARK: - Unit Conversion

    func convertUnit(to newUnit: String) {
        guard let currentValue = Double(value) else { return }

        let convertedValue: Double
        if newUnit == "mmol/L" && unit == "mg/dL" {
            // Convert mg/dL to mmol/L
            convertedValue = currentValue / 18.0
            value = String(format: "%.1f", convertedValue)
        } else if newUnit == "mg/dL" && unit == "mmol/L" {
            // Convert mmol/L to mg/dL
            convertedValue = currentValue * 18.0
            value = String(format: "%.0f", convertedValue)
        }

        // Validate with new unit
        if let doubleValue = Double(value) {
            valueError = validateValue(doubleValue)
        }
    }

    // MARK: - Save

    func saveReading() async {
        guard let glucoseValue = Double(value) else { return }

        isSaving = true
        saveError = nil

        do {
            guard let user = await authService.getCurrentUser(),
                  let patientId = user.linkedPatientId else {
                throw VitalError.noPatientId
            }

            // Convert to mg/dL for storage (standard FHIR unit)
            let valueInMgDl = unit == "mmol/L" ? glucoseValue * 18.0 : glucoseValue

            // Create glucose observation
            let observation = FHIRObservation(
                patientId: patientId,
                type: .bloodGlucose,
                effectiveDateTime: Date(),
                value: valueInMgDl,
                unit: "mg/dL",
                performerId: user.userId,
                performerType: .relatedPerson,
                note: mealTiming?.displayName
            )

            // Save to local store (adds to sync queue)
            try await localFHIRRepository.saveObservation(observation)

            // Play voice acknowledgement
            voicePlayer.playSuccess(type: .bloodGlucose)

            isSaving = false

        } catch {
            isSaving = false
            saveError = error.localizedDescription
            voicePlayer.playFailure()
        }
    }
}

File: android/app/src/main/java/com/carelog/ui/vitals/GlucoseViewModel.kt
package com.carelog.ui.vitals

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.carelog.auth.AuthRepository
import com.carelog.fhir.client.ObservationType
import com.carelog.fhir.models.FhirObservation
import com.carelog.fhir.repository.LocalFhirRepository
import com.carelog.voice.VoiceAcknowledgementPlayer
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.Instant
import java.util.UUID
import javax.inject.Inject

/**
 * ViewModel for glucose logging screen.
 */
@HiltViewModel
class GlucoseViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val localFhirRepository: LocalFhirRepository,
    private val voicePlayer: VoiceAcknowledgementPlayer
) : ViewModel() {

    private val _uiState = MutableStateFlow(GlucoseUiState())
    val uiState: StateFlow<GlucoseUiState> = _uiState.asStateFlow()

    fun updateValue(value: String) {
        _uiState.update { state ->
            val doubleValue = value.toDoubleOrNull()
            val error = validateValue(doubleValue, state.unit)
            state.copy(
                value = value,
                valueError = error,
                canSave = error == null && doubleValue != null
            )
        }
    }

    fun updateUnit(unit: String) {
        _uiState.update { state ->
            // Convert value if switching units
            val currentValue = state.value.toDoubleOrNull()
            val newValue = if (currentValue != null) {
                if (unit == "mmol/L" && state.unit == "mg/dL") {
                    // Convert mg/dL to mmol/L
                    String.format("%.1f", currentValue / 18.0)
                } else if (unit == "mg/dL" && state.unit == "mmol/L") {
                    // Convert mmol/L to mg/dL
                    String.format("%.0f", currentValue * 18.0)
                } else {
                    state.value
                }
            } else {
                ""
            }

            val doubleValue = newValue.toDoubleOrNull()
            val error = validateValue(doubleValue, unit)

            state.copy(
                unit = unit,
                value = newValue,
                valueError = error,
                canSave = error == null && doubleValue != null
            )
        }
    }

    fun updateMealTiming(timing: MealTiming) {
        _uiState.update { it.copy(mealTiming = timing) }
    }

    private fun validateValue(value: Double?, unit: String): String? {
        if (value == null) return null

        return when (unit) {
            "mg/dL" -> when {
                value < 20 -> "Too low (min 20)"
                value > 600 -> "Too high (max 600)"
                else -> null
            }
            "mmol/L" -> when {
                value < 1.1 -> "Too low (min 1.1)"
                value > 33.3 -> "Too high (max 33.3)"
                else -> null
            }
            else -> null
        }
    }

    fun saveReading(onSuccess: () -> Unit) {
        val state = _uiState.value
        val value = state.value.toDoubleOrNull() ?: return

        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, saveError = null) }

            try {
                val user = authRepository.getCurrentUser()
                val patientId = user?.linkedPatientId ?: throw IllegalStateException("No patient ID")

                // Convert to mg/dL for storage (standard FHIR unit)
                val valueInMgDl = if (state.unit == "mmol/L") {
                    value * 18.0
                } else {
                    value
                }

                // Create glucose observation
                val observation = FhirObservation(
                    id = UUID.randomUUID().toString(),
                    patientId = patientId,
                    type = ObservationType.BLOOD_GLUCOSE,
                    value = valueInMgDl,
                    unit = "mg/dL",
                    effectiveDateTime = Instant.now().toString(),
                    performerId = user?.userId,
                    note = state.mealTiming?.displayName
                )

                // Save to local store (adds to sync queue)
                localFhirRepository.saveObservation(observation)

                // Play voice acknowledgement
                try { voicePlayer.playSuccess(ObservationType.BLOOD_GLUCOSE) } catch (_: Exception) {}

                _uiState.update { it.copy(isSaving = false, saved = true) }

            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        saveError = e.message ?: "Failed to save"
                    )
                }
                voicePlayer.playFailure()
            }
        }
    }
}

/**
 * UI state for glucose screen.
 */
data class GlucoseUiState(
    val value: String = "",
    val unit: String = "mg/dL",
    val mealTiming: MealTiming? = null,
    val valueError: String? = null,
    val canSave: Boolean = false,
    val isSaving: Boolean = false,
    val saved: Boolean = false,
    val saveError: String? = null
)
